File: library/src/main/java/com/raizlabs/android/dbflow/sql/language/IndexedBy.java
package com.raizlabs.android.dbflow.sql.language;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.builder.Condition;
import com.raizlabs.android.dbflow.sql.builder.ConditionQueryBuilder;
import com.raizlabs.android.dbflow.structure.Model;

/**
 * Description: The INDEXED BY part of a SELECT/UPDATE/DELETE
 */
public class IndexedBy<ModelClass extends Model> implements WhereBase<ModelClass> {

    private final String mIndexName;

    private final WhereBase<ModelClass> mWhereBase;

    /**
     * Creates the INDEXED BY part of the clause.
     *
     * @param mIndexName The name of the index
     * @param whereBase  The base piece of this query
     */
    IndexedBy(@NonNull String mIndexName, WhereBase<ModelClass> whereBase) {
        this.mIndexName = mIndexName;
        this.mWhereBase = whereBase;
    }

    /**
     * @return a WHERE piece of this query
     */
    public Where<ModelClass> where() {
        return new Where<>(this);
    }

    /**
     * @param where The string part of where
     * @return a WHERE query with the specified string
     */
    public Where<ModelClass> where(String where) {
        return where().whereClause(where);
    }

    /**
     * @param conditionQueryBuilder The set of conditions used to build a WHERE query.
     * @return a WHERE query with the specified {@link com.raizlabs.android.dbflow.sql.builder.ConditionQueryBuilder}
     */
    public Where<ModelClass> where(ConditionQueryBuilder<ModelClass> conditionQueryBuilder) {
        return where().whereQuery(conditionQueryBuilder);
    }

    /**
     * @param conditions the list of conditions used to build a WHERE query.
     * @return a WHERE query with the specified {@link com.raizlabs.android.dbflow.sql.builder.Condition}
     */
    public Where<ModelClass> where(Condition... conditions) {
        return where().andThese(conditions);
    }

    @Override
    public Class<ModelClass> getTable() {
        return mWhereBase.getTable();
    }

    @Override
    public Query getQueryBuilderBase() {
        return mWhereBase.getQueryBuilderBase();
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder = new QueryBuilder(mWhereBase.getQuery())
                .append("INDEXED BY ").appendQuoted(mIndexName).appendSpace();
        return queryBuilder.getQuery();
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/query/IndexedBy.kt
package com.raizlabs.android.dbflow.query

import com.raizlabs.android.dbflow.quoteIfNeeded
import com.raizlabs.android.dbflow.sql.Query
import com.raizlabs.android.dbflow.query.property.IndexProperty
import com.raizlabs.android.dbflow.structure.ChangeAction

/**
 * Description: The INDEXED BY part of a SELECT/UPDATE/DELETE
 */
class IndexedBy<TModel : Any>
/**
 * Creates the INDEXED BY part of the clause.
 *
 * @param indexProperty The index property generated.
 * @param whereBase     The base piece of this query
 */
(private val indexProperty: IndexProperty<TModel>,
 private val whereBase: WhereBase<TModel>)
    : BaseTransformable<TModel>(whereBase.databaseWrapper, whereBase.table) {

    override val queryBuilderBase: Query
        get() = whereBase.queryBuilderBase

    override val query: String
        get() = buildString {
            append(whereBase.query)
            append(" INDEXED BY ")
            append(indexProperty.indexName.quoteIfNeeded())
            append(" ")
        }

    override val primaryAction: ChangeAction
        get() = whereBase.primaryAction
}
